src/main/java/org/jenkinsci/plugins/workflow/cps/CpsBodyExecution.java

```diff
@@ -194,15 +195,21 @@ private Continuable createContinuable(CpsThread currentThread, CpsCallableInvoca
         t.getExecution().runInCpsVmThread(new FutureCallback<CpsThreadGroup>() {
             @Override public void onSuccess(CpsThreadGroup g) {
                 try {
+                    if (g.threads.size() == 1) {
+                        result.set(Collections.singletonList(g.threads.firstEntry().getValue().getStep()));
```

Comment:
First of all, you are failing to check for a null return value from `getStep`. So you might be returning `[null]`, which is illegal.

That aside, I doubt this optimization is correct. You are making no effort to verify that the thread’s `head` is in fact a child of this body’s start node. Now if the program in fact has only one active thread, then its single head should be a child of any _active_ body execution. But there is no documented guarantee that this method can be called only on an active body.

This method seems like a good candidate for consuming the new API rather than doing a linear scan, so I am not sure why you are not doing so.